Starting a branch for read and write features

zostay opened the pull request:
This is work in progress, but I wanted to submit a pull request to get your feedback.

So far I revamped internals and added support for doing basic reads. Writes work as
before (with new internals).

zostay committed:
Starting a branch for read and write features

So far revamped internals and adds support for doing basic reads. Writes work as
before.

hoelzro commented on lib/IO/String.pm:
```diff
@@ -36,10 +36,48 @@ TODO
+
+    multi method open(IO::String:D:) {
+        my Str $new-buffer = '';
+        $!buffer := $new-buffer;
```
`$new-buffer` is redundant, isn't it? Can't we just assign `''` to `$!buffer` here (and in the other variant) below?

zostay replied:
Possibly. I need to play with it a bit to see if I understand how binding works. If we eliminate binding, then we don't need to worry about it.

hoelzro commented on lib/IO/String.pm:
```diff
@@ -36,10 +36,48 @@ TODO
+    }
+
+    multi method open(IO::String:D: Str $buffer is rw) {
+        $!buffer := $buffer;
```
This "read-write" buffer being passed in kind of bugs me a bit;  if I understand this correctly, that means that this code will alter `$s`, right?

``` perl6
my $s = "foobar\n";
my $io = IO::String.open($s);
$io.say: "hello";
```

zostay replied:
Yes. This is a feature the Perl 5 version of IO::String has and I think it's reasonable. On the other hand, it might be surprising, so maybe it should be available with a flag instead. Like:

``` perl6
multi method open(IO::String:D: Str $buffer is rw, :bind!) { ... }
multi method open(IO::String:D: Str $buffer) { ... }
```

hoelzro replied:
I'm totally good with the feature if we have an explicit parameter like `:$bind`!

hoelzro commented on lib/IO/String.pm:
```diff
@@ -36,10 +36,48 @@ TODO
+            return $.buffer.substr($start);
+        }
+
+        my $len = 1 + $next-nl.key - $start;
```
I have a feeling this won't work for `$.nl-in` values longer than 1 character; maybe this should be `my $len = $next-nl.value.chars + $next-nl.key - $start`?

hoelzro commented on t/03-get.t:
```diff
@@ -0,0 +1,11 @@
+is $s.get, "hello,", "got first line";
+is $s.get, "world!", "got second line";
+ok !$s.get, "got nothing for third line";
+ok $s.eof, "we have eof";
```
Could we add tests for `nl-in` values with more than 1 character and an `.eof` check before we've run out of characters?

hoelzro commented:
@zostay Thanks for working on this!  I've left some comments on the code; please have a look when you have time.

zostay commented:
I hope to have a chance to look at this after work later. The comments are very helpful.

hoelzro commented:
@zostay Excellent, keep me posted.  Thanks again for helping to develop IO::String!

zostay committed:
Making binding manual with :bind arg

zostay committed:
Adding a test for multi-char newline

Want to make sure the code still works if multi-char nl-in is used.

zostay committed:
Adding test for not yet eof

zostay committed:
docs, test lines, test without chomp

zostay committed:
more documentation

zostay commented:
I think I addressed each of the suggestions you made. I also added more documentation and tests. I still need to look into what other methods should be implemented to complete the implementation.

hoelzro commented:
This looks pretty good; I'll merge later today!

zostay committed:
added support for readchars()

zostay committed:
Add tell

zostay committed:
Add flush as a no-op, just in case something tries to call it

zostay commented:
I added one last push and I am done now. This adds readchars(), seek(), and tell() so you can do character counting. I'm not going to implement slurp-rest() and read(), etc. because that involves encoding support.

Writes lack support for seek() and tell(), so print() does not overwrite as it should and put() and friends. That needs to be added to complete read/write support.

zostay committed:
Adding getc

zostay committed:
Adding better support for print/seek

Also lets you seek to the end from the start by specifying :a to open.

zostay committed:
Adding support for put

zostay committed:
Adding in .gist and .perl.

zostay commented:
So, I broke my promise. I went ahead and fixed print. This adds a `:a` option to open, makes seek work with print, also adds getc, which I overlooked, adds put, and adds gist and perl methods. It adds more tests for many of these things.

Most of what remains is related to encoding, which I haven't dived into. It wouldn't be hard to add, though. Mostly, it's just a matter of making sure .encode and .decode are called in time to count bytes rather than chars or to write bytes to the string. An IO::Buf might be a nice companion to this object.

hoelzro commented:
One quick thought: `getc` returns `Nil` at end of stream; should `readchars` return `Nil` as well?

hoelzro commented:
I'm going to merge to get the ball rolling, but `IO::Handle` appears to return `Str` on end of stream.

zostay commented:
Thanks for making this happen and sorry I didn't get back to you. It has been a crazy couple of days for me. Anyway, you rock! :+1:

Cheers.
